README.md
# Internet Provider Monitor

## What

This project has a simple goal of monitoring the internet provider performance: lantency and download/upload speed. Also, it is an opportunity to connect multiple technologies together and deliver an observability framework comprising of alerting, monitoring and logging (tracing will come later).

## Why

Of course you could run any speed test, or ping to detect if you network is down, but how fun is it? Also, this project will give you historical performance data for analysis.

## How

A few technologies were used:

- The core backend application (Internet Monitor - IM) was developed using Python with a job scheduler, instrumented with Prometheus metrics, that triggers jobs to test ICMP echo request latency average RTT, download/upload performance
- Prometheus /metrics endpoint was added to the code and a few custom metrics to track latency, and download/upload speed (MB/s)
- Prometheus server to scrape the data from IM
- Grafana is used to present the data and show the historical information
- Python standard logging with details about the job execution and the overall app messages
- Docker and docker-compose is used to build and run the monitor app, prometheus, grafana and alert manager
- Flake8 to check for pep8 issues
- mypy to check for type errors
- bandit to check for vulnerabilities
- pytest for unit testing
- black to format our python code
- GNU Make to facilite our development process (i.e. build, test, etc)
- cAdvisor to expose container information such as memory/cpu consumption for all containers running under docker-compose control
- Loki to show logs in grafana
- Node exporter to gather node metrics
- Alertmanager to manager alerts sent from prometheus

## Tech stack

- Python 3
- Docker
- docker-compose
- prometheus
- alertmanager
- grafana
- cAdvisor
- GNU Make
- node-exporter
- loki

### Python major libraries and tools

- flake8
- prometheus_client
- bandit
- requests
- mypy
- coverage
- pytest
- icmplib
- isort

## Running the code

### Pre-requisites

- A working docker and docker-compose
- Git to download the code

### How to run

```bash
git clone https://github.com/global/internet-monitor.git
make init
make build
make run
make urls
```

Open:

- <http://localhost:9090/> - to access prometheus
- <http://localhost:3000/> - to access grafana (login: admin / pass: admin)
  - There is a pre-configured dashboard called `Internet Monitoring`. When logged in, click to access the information
  - To visualize all the data, wait about 5 min, as this is the time the download speed job kicks in
- <http://localhost:8000/metrics> - to access the internet-monitor metrics
- <http://localhost:8080/> - to access cAdvisor (you can also access the metrics through prometheus and grafana)

This is the `Internet Monitoring` you will get when opening grafana:

![Grafana Internet Monitor Dashboard](docs/grafana.png "Grafana Internet Monitor Dashboard")

## Development

## Bugs and Known Issues

- Sometimes my docker for mac stop accepting inbound ICMP traffic and I need to restart it to get  `ping` working again.

## TODO

- Create a demo video
- Integrate our alerthook with alertmanager
- Add container resources dashboard into grafana
- Integrate alertmanager with external notification services (i.e. Pagerduty, Xmatters, etc.)

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.4.2]

### Changed

- Removed unused github action
- Applied python code formatting to IM
- Changed docker-compose to docker compose command in the Makefile

## [0.4.1]

### Changed

- Fixed loki installation error in Makefile

## [0.4.0]

### Added

- Unit testing added with 100% coverage and check to IM
- Docstrings were added to the code in all methods/modules/packages to IM
- Introduced type hints checks with mypi, but still requires code refactoring
- Added automatic formatting using black and sorting python imports using isort
  to IM
- Added setup.py and moved code to im subfolder

### Changed

- CHANGELOG format changed to use keepachangelog.com format
- Internet Monitor (IM) code refactored, but maintained the same functionality

## [0.3.0]

### Added

- Added upload speed calculation
- Added labels to the Dockerfile
- Added make support to facilitate build/run
- Added support for Pipenv and python 3
- Added more widgets into the dashboard

### Changed

- Jobs now starts almost immediately after the program starts
- Alerthook disabled until we can finish its integration with alertmanager
- Fixed metrics names and values

## [0.2.0]

### Added

- Added alertmanager, node-exporter and a webhook to receive alerts
- Added a CHANGELOG.md file

### Changed

- Upgrade of grafana to 7.2 version

## [0.1.0]

### Added

- Initial code base for internet monitor
- Grafana, prometheus, loki for logs and cAdvisor are integrated


im/tests/test_monitor.py
import json
import random
import sys
from importlib import reload
from unittest.mock import Mock, patch

import pytest
from prometheus_client import REGISTRY

from im import monitor


@pytest.fixture
def icmp_mocked_host():
    mock = Mock()
    mock.avg_rtt = random.randint(1, 10)
    mock.packet_loss = random.randint(0, 10)
    mock.max_rtt = random.randint(1, 5)
    mock.min_rtt = random.randint(5, 10)

    return mock


def clean_prometheus_registry():
    collectors = list(REGISTRY._collector_to_names.keys())
    for collector in collectors:
        REGISTRY.unregister(collector)


def test_latency(icmp_mocked_host):
    """
    Test if ping succeeds and generates the right metrics.
    """

    clean_prometheus_registry()
    reload(monitor)

    before_up = REGISTRY.get_sample_value("internet_monitor_up")
    before_ping_total = REGISTRY.get_sample_value("internet_monitor_ping_total")
    before_ping_packet_loss_total = REGISTRY.get_sample_value(
        "internet_monitor_ping_packet_loss_total"
    )
    before_ping_jitter_seconds = REGISTRY.get_sample_value(
        "internet_monitor_ping_jitter_seconds"
    )
    before_ping_latency_seconds = REGISTRY.get_sample_value(
        "internet_monitor_ping_latency_seconds"
    )

    assert before_up == 0
    assert before_ping_total == 0
    assert before_ping_packet_loss_total == 0
    assert before_ping_jitter_seconds == 0

    with patch("im.monitor.ping") as mock_ping:
        mock_ping.return_value = icmp_mocked_host
        monitor.latency("1.1.1.1")

    after_up = REGISTRY.get_sample_value("internet_monitor_up")
    after_ping_total = REGISTRY.get_sample_value("internet_monitor_ping_total")
    after_ping_packet_loss_total = REGISTRY.get_sample_value(
        "internet_monitor_ping_packet_loss_total"
    )
    after_ping_jitter_seconds = REGISTRY.get_sample_value(
        "internet_monitor_ping_jitter_seconds"
    )
    after_ping_latency_seconds = REGISTRY.get_sample_value(
        "internet_monitor_ping_latency_seconds_sum"
    )

    assert after_up == 1
    assert after_ping_total == 1
    assert after_ping_packet_loss_total == icmp_mocked_host.packet_loss
    assert (
        after_ping_jitter_seconds
        == (icmp_mocked_host.max_rtt - icmp_mocked_host.min_rtt) / 1000
    )
    assert after_ping_latency_seconds == icmp_mocked_host.avg_rtt / 1000


def test_latency_failure(icmp_mocked_host):
    """
    Test if ping fails and generates the right metrics.
    """

    clean_prometheus_registry()
    reload(monitor)

    before_up = REGISTRY.get_sample_value("internet_monitor_up")
    before_ping_total = REGISTRY.get_sample_value("internet_monitor_ping_total")
    before_ping_failures_total = REGISTRY.get_sample_value(
        "internet_monitor_ping_failures_total"
    )

    assert before_up == 0
    assert before_ping_total == 0
    assert before_ping_failures_total == 0

    with patch("im.monitor.ping") as mock_ping:
        mock_ping.side_effect = Exception("boom!")
        monitor.latency("1.1.1.1")

    after_up = REGISTRY.get_sample_value("internet_monitor_up")
    after_ping_total = REGISTRY.get_sample_value("internet_monitor_ping_total")
    before_ping_failures_total = REGISTRY.get_sample_value(
        "internet_monitor_ping_failures_total"
    )

    assert after_up == 0
    assert after_ping_total == 1
    assert before_ping_failures_total == 1


@pytest.fixture
def http_mocked_response():
    mock = Mock()
    mock.content = "ok"
    return mock


def test_download_speed(http_mocked_response):
    """
    Test if download succeeds and calculates the appropriate size and successfull job execution.
    """

    clean_prometheus_registry()
    reload(monitor)

    before_download_duration_seconds_sum = REGISTRY.get_sample_value(
        "internet_monitor_download_duration_seconds_sum"
    )
    before_download_size_bytes = REGISTRY.get_sample_value(
        "internet_monitor_download_size_bytes"
    )
    before_download_total = REGISTRY.get_sample_value("internet_monitor_download_total")

    assert before_download_duration_seconds_sum == 0
    assert before_download_size_bytes == 0
    assert before_download_total == 0

    with patch("im.monitor.requests") as mocked_requests:
        mocked_requests.get.return_value = http_mocked_response
        monitor.download_speed("http://localhost")

    after_download_duration_seconds_sum = REGISTRY.get_sample_value(
        "internet_monitor_download_duration_seconds_sum"
    )
    after_download_size_bytes = REGISTRY.get_sample_value(
        "internet_monitor_download_size_bytes"
    )
    after_download_total = REGISTRY.get_sample_value("internet_monitor_download_total")

    assert isinstance(after_download_duration_seconds_sum, float)
    assert after_download_size_bytes == len(http_mocked_response.content)
    assert after_download_total == 1


def test_download_speed_failures(http_mocked_response):
    """
    Test if download fails and failure metrics are updated accordingly.
    """

    clean_prometheus_registry()
    reload(monitor)

    before_download_duration_seconds_sum = REGISTRY.get_sample_value(
        "internet_monitor_download_duration_seconds_sum"
    )
    before_download_failures_total = REGISTRY.get_sample_value(
        "internet_monitor_download_failures_total"
    )
    before_download_total = REGISTRY.get_sample_value("internet_monitor_download_total")

    assert before_download_duration_seconds_sum == 0
    assert before_download_failures_total == 0
    assert before_download_total == 0

    with patch("im.monitor.requests") as mocked_requests:
        mocked_requests.get.side_effect = Exception("boom!")
        monitor.download_speed("http://localhost")

    after_download_duration_seconds_sum = REGISTRY.get_sample_value(
        "internet_monitor_download_duration_seconds_sum"
    )
    after_download_failures_total = REGISTRY.get_sample_value(
        "internet_monitor_download_failures_total"
    )
    after_download_total = REGISTRY.get_sample_value("internet_monitor_download_total")

    assert isinstance(after_download_duration_seconds_sum, float)
    assert after_download_failures_total == 1
    assert after_download_total == 1


def test_upload_speed(http_mocked_response):
    """
    Test if upload succeeds and calculates the appropriate size and successfull job execution.
    """

    clean_prometheus_registry()
    reload(monitor)

    before_upload_duration_seconds_sum = REGISTRY.get_sample_value(
        "internet_monitor_upload_duration_seconds_sum"
    )
    before_upload_size_bytes = REGISTRY.get_sample_value(
        "internet_monitor_upload_size_bytes"
    )
    before_upload_total = REGISTRY.get_sample_value("internet_monitor_upload_total")

    assert before_upload_duration_seconds_sum == 0
    assert before_upload_size_bytes == 0
    assert before_upload_total == 0

    with patch("im.monitor.requests") as mocked_requests:
        with patch("im.monitor.os.stat") as mocked_stat:
            mocked_requests.post.return_value = http_mocked_response
            mocked_stat.return_value.st_size = len(http_mocked_response.content)
            monitor.upload_speed("http://localhost")

    after_upload_duration_seconds_sum = REGISTRY.get_sample_value(
        "internet_monitor_upload_duration_seconds_sum"
    )
    after_upload_size_bytes = REGISTRY.get_sample_value(
        "internet_monitor_upload_size_bytes"
    )
    after_upload_total = REGISTRY.get_sample_value("internet_monitor_upload_total")

    assert isinstance(after_upload_duration_seconds_sum, float)
    assert after_upload_size_bytes == len(http_mocked_response.content)
    assert after_upload_total == 1


def test_upload_speed_failures(http_mocked_response):
    """
    Test if upload fails and failure metrics are updated accordingly.
    """

    clean_prometheus_registry()
    reload(monitor)

    before_upload_duration_seconds_sum = REGISTRY.get_sample_value(
        "internet_monitor_upload_duration_seconds_sum"
    )
    before_upload_failures_total = REGISTRY.get_sample_value(
        "internet_monitor_upload_failures_total"
    )
    before_upload_total = REGISTRY.get_sample_value("internet_monitor_upload_total")

    assert before_upload_duration_seconds_sum == 0
    assert before_upload_failures_total == 0
    assert before_upload_total == 0

    with patch("im.monitor.requests") as mocked_requests:
        mocked_requests.post.side_effect = Exception("boom!")
        monitor.upload_speed("http://localhost")

    after_upload_duration_seconds_sum = REGISTRY.get_sample_value(
        "internet_monitor_upload_duration_seconds_sum"
    )
    after_upload_failures_total = REGISTRY.get_sample_value(
        "internet_monitor_upload_failures_total"
    )
    after_upload_total = REGISTRY.get_sample_value("internet_monitor_upload_total")

    assert isinstance(after_upload_duration_seconds_sum, float)
    assert after_upload_failures_total == 1
    assert after_upload_total == 1


im/main.py
#!/usr/bin/env python3
"""
This app monitors your internet connection and generates prometheus metrics.
"""
import sys
import os
import logging
import requests
import yaml
import datetime

from pytz import utc
from apscheduler.schedulers.background import BlockingScheduler
from apscheduler.executors.pool import ThreadPoolExecutor
from prometheus_client import start_http_server, Counter, Gauge, Summary, Histogram
from prometheus_client.utils import INF
from icmplib import ping

# ping metrics
PING_REQUESTS = Counter(
    "internet_monitor_ping_total", "Total ping requests made to 1.1.1.1"
)
PING_FAILURES = Counter(
    "internet_monitor_ping_failures_total", "Total ping requests failed made to 1.1.1.1"
)
PING_PACKET_LOSS = Counter(
    "internet_monitor_ping_packet_loss_total",
    "Number of packets lost while checking latency",
)
PING_JITTER = Gauge("internet_monitor_ping_jitter_seconds", "ICMP Jitter")
UP = Gauge("internet_monitor_up", "Internet is up or down")
PING_LATENCY = Summary(
    "internet_monitor_ping_latency_seconds", "Ping latency to 1.1.1.1"
)

# download metrics
DOWNLOAD_DURATION = Histogram(
    "internet_monitor_download_duration_seconds",
    "Download latency",
    buckets=(1, 2, 5, 7, 10, 15, 20, 50, 100, INF),
)
DOWNLOAD_REQUEST_SIZE = Gauge(
    "internet_monitor_download_size_bytes", "Bytes downloaded"
)
DOWNLOAD_REQUESTS = Counter(
    "internet_monitor_download_total", "Number of times the download job runs"
)
DOWNLOAD_FAILURES = Counter(
    "internet_monitor_download_failures_total", "Number of times the download job fails"
)

# upload metrics
UPLOAD_DURATION = Histogram(
    "internet_monitor_upload_duration_seconds",
    "Upload latency",
    buckets=(1, 2, 5, 7, 10, 15, 20, 50, 100, INF),
)
UPLOAD_REQUESTS = Counter(
    "internet_monitor_upload_total", "Number of times the upload job runs"
)
UPLOAD_FAILURES = Counter(
    "internet_monitor_upload_failures_total", "Number of times the upload job fails"
)
UPLOAD_REQUEST_SIZE = Gauge("internet_monitor_upload_size_bytes", "Bytes uploaded")

LOGGER = logging.getLogger("internet.monitor")


def latency(dest):
    """
    Calculate the average rtt time using ICMP echo request
    """

    PING_REQUESTS.inc()
    try:
        with PING_FAILURES.count_exceptions():
            host = ping(dest, count=2, interval=0.5)
        PING_LATENCY.observe(host.avg_rtt / 1000)
        PING_PACKET_LOSS.inc(host.packet_loss)
        PING_JITTER.set((host.max_rtt - host.min_rtt) / 1000)
        UP.set(1)
    except Exception as e:
        LOGGER.error("Could not process ICMP Echo requests.")
        print(e)
        UP.set(0)


@DOWNLOAD_DURATION.time()
def download_speed(url):
    """
    Calculates the download speed based request size and the time it takes
    """
    DOWNLOAD_REQUESTS.inc()

    try:
        resp = requests.get(url)
        DOWNLOAD_REQUEST_SIZE.set(len(resp.content))

    except Exception:
        LOGGER.error("Cannot download the file %s", url)
        DOWNLOAD_FAILURES.inc()


@UPLOAD_DURATION.time()
def upload_speed():
    """
    Calculates the upload speed based on http POST method
    """
    UPLOAD_REQUESTS.inc()
    try:
        files = {"file": open("test-upload", "rb")}
        requests.post("https://file.io/?expires=1d", files=files)
        UPLOAD_REQUEST_SIZE.set(os.stat("test-upload").st_size)
    except Exception:
        LOGGER.error("Cannot upload the file")
        UPLOAD_FAILURES.inc()


def load_configuration(filename):
    """
    Load YAML configuration file and return a dict
    """
    with open(filename) as conf:
        return yaml.load(conf, Loader=yaml.FullLoader)


def main(config):
    """
    Setup logging, start the job scheduler and serve prometheus metrics
    """

    LOGGER.info("Sarting application at http://localhost:8000")

    executors = {
        "default": ThreadPoolExecutor(5),
    }
    job_defaults = {"coalesce": False, "max_instances": 3}

    scheduler = BlockingScheduler(
        executors=executors, job_defaults=job_defaults, timezone=utc
    )

    start_job_date = datetime.datetime.now() - datetime.timedelta(seconds=590)
    scheduler.add_job(
        download_speed,
        "interval",
        seconds=600,
        args=[config["downloadURL"]],
        id="download_speed",
        start_date=start_job_date,
    )

    start_job_date = datetime.datetime.now() - datetime.timedelta(seconds=50)
    scheduler.add_job(
        latency,
        "interval",
        seconds=60,
        args=[config["icmpDestHost"]],
        id="ping",
        start_date=start_job_date,
    )

    # create temporary upload file
    with open("test-upload", "wb") as out:
        out.truncate(1024 * 1024 * 50)

    start_job_date = datetime.datetime.now() - datetime.timedelta(seconds=3590)
    scheduler.add_job(
        upload_speed,
        "interval",
        seconds=3600,
        id="upload_speed",
        start_date=start_job_date,
    )

    # start prometheus server to serve /metrics and /describe endpoints
    start_http_server(addr="0.0.0.0", port=8000)
    scheduler.start()


if __name__ == "__main__":

    # load app configuration
    CONFIG = load_configuration("monitor.yml")

    # Setting up logging
    LOGGER.setLevel(CONFIG["logLevel"])
    # create console handler with a higher log level
    CH = logging.StreamHandler()
    CH.setLevel(CONFIG["logLevel"])
    # create formatter and add it to the handlers
    FORMATTER = logging.Formatter(
        "%(asctime)s - %(name)s - %(levelname)s - %(message)s"
    )
    CH.setFormatter(FORMATTER)
    # add the handlers to the LOGGER and apscheduler LOGGER
    LOGGER.addHandler(CH)
    logging.getLogger("apscheduler.executors.default").addHandler(CH)
    logging.getLogger("apscheduler.executors.default").setLevel(CONFIG["logLevel"])

    try:
        main(CONFIG)
    except KeyboardInterrupt:
        LOGGER.info("Shutting down internet-monitor...")
        sys.exit(0)


im/im/monitor.py
# -*- coding: utf-8 -*-
"""
This app monitors your internet connection and generates prometheus metrics.
"""
import datetime
import logging
import os
import sys

import requests
from apscheduler.executors.pool import ThreadPoolExecutor
from apscheduler.schedulers.background import BlockingScheduler
from icmplib import ping
from prometheus_client import Counter, Gauge, Histogram, Summary, start_http_server
from prometheus_client.utils import INF
from pytz import utc

from im.config import CONFIG

# ping metrics
PING_REQUESTS = Counter(
    "internet_monitor_ping_total", "Total ping requests made to 1.1.1.1"
)
PING_FAILURES = Counter(
    "internet_monitor_ping_failures_total", "Total ping requests failed made to 1.1.1.1"
)
PING_PACKET_LOSS = Counter(
    "internet_monitor_ping_packet_loss_total",
    "Number of packets lost while checking latency",
)
PING_JITTER = Gauge("internet_monitor_ping_jitter_seconds", "ICMP Jitter")
UP = Gauge("internet_monitor_up", "Internet is up or down")
PING_LATENCY = Summary(
    "internet_monitor_ping_latency_seconds", "Ping latency to 1.1.1.1"
)

# download metrics
DOWNLOAD_DURATION = Histogram(
    "internet_monitor_download_duration_seconds",
    "Download latency",
    buckets=(1, 2, 5, 7, 10, 15, 20, 50, 100, INF),
)
DOWNLOAD_REQUEST_SIZE = Gauge(
    "internet_monitor_download_size_bytes", "Bytes downloaded"
)
DOWNLOAD_REQUESTS = Counter(
    "internet_monitor_download_total", "Number of times the download job runs"
)
DOWNLOAD_FAILURES = Counter(
    "internet_monitor_download_failures_total", "Number of times the download job fails"
)

# upload metrics
UPLOAD_DURATION = Histogram(
    "internet_monitor_upload_duration_seconds",
    "Upload latency",
    buckets=(1, 2, 5, 7, 10, 15, 20, 50, 100, INF),
)
UPLOAD_REQUESTS = Counter(
    "internet_monitor_upload_total", "Number of times the upload job runs"
)
UPLOAD_FAILURES = Counter(
    "internet_monitor_upload_failures_total", "Number of times the upload job fails"
)
UPLOAD_REQUEST_SIZE = Gauge("internet_monitor_upload_size_bytes", "Bytes uploaded")

LOGGER = logging.getLogger("internet.monitor")


def latency(dest):
    """Calculate the internet latency stats based on ICMP.

    This method uses icmp echo request/response and calculate the average rtt, packet loss and jitter.

    Args:
        dest (string): hostname or IP address

    """

    PING_REQUESTS.inc()
    try:
        with PING_FAILURES.count_exceptions():
            LOGGER.debug('PING - destination: ' + str(dest))
            host = ping(str(dest), count=2, interval=0.5)
        PING_LATENCY.observe(host.avg_rtt / 1000)
        PING_PACKET_LOSS.inc(host.packet_loss)
        PING_JITTER.set((host.max_rtt - host.min_rtt) / 1000)
        LOGGER.debug('PING - max_rtt: ' + str(host.max_rtt))
        LOGGER.debug('PING - min_rtt: ' + str(host.min_rtt))
        LOGGER.debug('PING - avg_rtt: ' + str(host.avg_rtt))
        LOGGER.debug('PING - packet_loss: ' + str(host.packet_loss))
        UP.set(1)
    except Exception as e:
        LOGGER.error("Could not process ICMP Echo requests: %s", e)
        UP.set(0)


@DOWNLOAD_DURATION.time()
def download_speed(url):
    """
    Calculates the download speed based request size and the time it takes
    """
    DOWNLOAD_REQUESTS.inc()

    try:
        LOGGER.info("downloading a file...")
        resp = requests.get(url)
        DOWNLOAD_REQUEST_SIZE.set(len(resp.content))
    except Exception:
        LOGGER.error("Cannot download the file %s", url)
        DOWNLOAD_FAILURES.inc()


@UPLOAD_DURATION.time()
def upload_speed(url):
    """
    Calculates the upload speed based on http POST method
    """
    UPLOAD_REQUESTS.inc()
    try:
        files = {"file": open("test-upload", "rb")}
        requests.post(url, files=files)
        UPLOAD_REQUEST_SIZE.set(os.stat("test-upload").st_size)
    except Exception:
        LOGGER.error("Cannot upload the file")
        UPLOAD_FAILURES.inc()


def job_scheduler(config):  # pragma: no cover
    """
    Setup logging, start the job scheduler and serve prometheus metrics
    """

    LOGGER.info("Sarting application at http://localhost:8000")

    executors = {
        "default": ThreadPoolExecutor(5),
    }
    job_defaults = {"coalesce": False, "max_instances": 5}

    scheduler = BlockingScheduler(
        executors=executors, job_defaults=job_defaults, timezone=utc
    )

    scheduler.add_job(
        func=download_speed,
        trigger="interval",
        max_instances=1,
        seconds=config["jobs"]["download"]["interval"],
        args=[config["downloadURL"]],
        id="download_speed",
        next_run_time=datetime.datetime.utcnow(),
        start_date=datetime.datetime.utcnow(),
    )

    scheduler.add_job(
        func=latency,
        trigger="interval",
        seconds=config["jobs"]["ping"]["interval"],
        args=[config["icmpDestHost"]],
        id="ping",
        next_run_time=datetime.datetime.utcnow(),
        start_date=datetime.datetime.utcnow(),
    )

    # create temporary upload file
    with open("test-upload", "wb") as out:
        out.truncate(1024 * 1024 * 50)

    scheduler.add_job(
        func=upload_speed,
        trigger="interval",
        seconds=config["jobs"]["upload"]["interval"],
        args=[config["uploadURL"]],
        id="upload_speed",
        next_run_time=datetime.datetime.utcnow(),
        start_date=datetime.datetime.utcnow(),
    )

    # start prometheus server to serve /metrics and /describe endpoints
    start_http_server(addr="0.0.0.0", port=8000)  # nosec
    scheduler.start()


def main():  # pragma: no cover
    """
    Start up scheduler and logging facilities.
    """

    # load app configuration

    # Setting up logging
    LOGGER.setLevel(CONFIG["logLevel"])
    # create console handler with a higher log level
    CH = logging.StreamHandler()
    CH.setLevel(CONFIG["logLevel"])
    # create formatter and add it to the handlers
    FORMATTER = logging.Formatter(
        "%(asctime)s - %(name)s - %(levelname)s - %(message)s"
    )
    CH.setFormatter(FORMATTER)
    # add the handlers to the LOGGER and apscheduler LOGGER
    LOGGER.addHandler(CH)
    logging.getLogger("apscheduler.executors.default").addHandler(CH)
    logging.getLogger("apscheduler.executors.default").setLevel(CONFIG["logLevel"])

    LOGGER.debug("PYTHONPATH: %s", sys.path)
    try:
        job_scheduler(CONFIG)
    except KeyboardInterrupt:
        LOGGER.info("Shutting down internet-monitor...")
        sys.exit(0)


if __name__ == "__main__":  # pragma: no cover
    main()
